### PR TITLE
SC-6275 add etag handling to github-pull-requests

### PR DESCRIPTION
**Pull request by kissmikijr**
Adds a GithubPullRequestsContext, where we can store the etags and the data.
Make filtering happen on the frontend to not store all the possible combination of state. 

#### :heavy_check_mark: Checklist

- [ ] Added tests for new functionality and regression tests for bug fixes
- [ ] Screenshots of before and after attached (for UI changes)
- [x] Added or updated documentation (if applicable)

### CODE REVIEW

**Comment by kissmikijr** [comment]
[sc-6275]

File: plugins/frontend/backstage-plugin-github-pull-requests/src/api/GithubPullRequestsClient.ts [hunk 1]
```diff
@@ -36,28 +39,37 @@ export class GithubPullRequestsClient implements GithubPullRequestsApi {
     page?: number;
     state?: PullRequestState;
     baseUrl: string | undefined;
+    etag: string
   }): Promise<{
     maxTotalItems?: number;
     pullRequestsData: PullsListResponseData;
+    etag?: string
```
Comment (Xantier) [inline plugins/frontend/backstage-plugin-github-pull-requests/src/api/GithubPullRequestsClient.ts:46]: Is this not returned from Octokit every time?

File: plugins/frontend/backstage-plugin-github-pull-requests/src/components/usePullRequests.ts [hunk 2]
```diff
@@ -14,131 +14,130 @@
...
+        baseUrl,
+        etag: state && prState[state].etag || ""
+      })
+      if (maxTotalItems) {
+        setTotal(maxTotalItems);
+      }
+      if (etag) {
+        setPrState((current: PrState) => ({
+          ...current,
+          ...{ [state]: { ...current[state], etag } }
```
Comment (Xantier) [inline plugins/frontend/backstage-plugin-github-pull-requests/src/components/usePullRequests.ts:74]: Would this be the same written as:
```suggestion
          [state]: { ...current[state], etag }
```
?

File: plugins/frontend/backstage-plugin-github-pull-requests/src/components/GithubPullRequestsContext.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import React, { FC, useState, useContext } from "react"
+import { PrState, PullRequestsContextData } from "../types"
+
+const GithubPullRequestsContext = React.createContext<PullRequestsContextData>(
+    {
+        prState: { open: { etag: "", data: [] }, closed: { etag: "", data: [] }, all: { etag: "", data: [] } },
```
Comment (Xantier) [inline plugins/frontend/backstage-plugin-github-pull-requests/src/components/GithubPullRequestsContext.tsx:6]: Maybe create an `initialValue` variable that can be cloned for use in here and the useState hook call below. This is getting a bit complex so easier to keep a single item up to date that way.

**Review by Xantier — APPROVED** [review]

File: plugins/frontend/backstage-plugin-github-pull-requests/src/api/GithubPullRequestsClient.ts (hunk 1 above)
Comment (kissmikijr) [reply plugins/frontend/backstage-plugin-github-pull-requests/src/api/GithubPullRequestsClient.ts:46]: Oh, you are right, for pullrequests it is returned. The api does not return for every endpoint the etag, but in this case all I care is the pullrequests. Ty!

File: plugins/frontend/backstage-plugin-github-pull-requests/src/components/GithubPullRequestsContext.tsx (hunk 3 above)
Comment (kissmikijr) [reply plugins/frontend/backstage-plugin-github-pull-requests/src/components/GithubPullRequestsContext.tsx:6]: Yeah, done

**Review by Xantier — DISMISSED** [review]

**Review by eoghanmcilwaine — COMMENTED** [review]
Looks good. Zustand is a cool addition.

**Comment by kissmikijr** [comment]
Yea, I am trying it out with zustand, I didn't know it, but it uses the global state so it won't become a breaking change after all and we won't use the providers! I think it will be better with zustand

**Comment by kissmikijr** [comment]
[SC-6502]

**Pull request closed without merging**